sqlglot/optimizer/qualify_columns.py

```diff
@@ -213,6 +214,31 @@ def _qualify_columns(scope, resolver):
             # column_table can be a '' because bigquery unnest has no table alias
             if column_table:
                 column.set("table", column_table)
+        elif column_table not in scope.sources:
+            # structs are used like tables (e.g. "struct"."field"), so they need to be qualified
+            # separately by parsing them as <table>.<column>.dot.dot.dot ...
+
+            struct_fields = [val for val in reversed(list(column.args.values())) if val is not None]
+            struct_root = struct_fields[0]
+
+            if struct_root.name in scope.sources:
+                # struct is already qualified, but we still need to change the AST representation
+                struct_table = struct_root
+                struct_fields = struct_fields[1:]
+                struct_root = struct_fields[0]
+            else:
+                struct_table = resolver.get_table(struct_root.name)
```

Comment:
why would this case happen?

Reply:
This is executed if the selection isn't already qualified, e.g. `"struct"."f1"` instead of `"cte"."struct"."f1"`.